src/operator/contrib/csr_neighborhood_sample-inl.h

```diff
@@ -0,0 +1,281 @@
+  }
+};
+
+static bool CSRNeighborSampleStorageType(const nnvm::NodeAttrs& attrs,
+                                         const int dev_mask,
+                                         DispatchMode* dispatch_mode,
+                                         std::vector<int> *in_attrs,
+                                         std::vector<int> *out_attrs) {
+  CHECK_EQ(in_attrs->size(), 3);
+  CHECK_EQ(out_attrs->size(), 1);
+
+  CHECK_EQ(in_attrs->at(0), mxnet::kCSRStorage);
+  CHECK_EQ(in_attrs->at(1), mxnet::kDefaultStorage);
+  CHECK_EQ(in_attrs->at(2), mxnet::kDefaultStorage);
+
+  bool success = true;
+  if (!type_assign(&(*out_attrs)[0], mxnet::kDefaultStorage)) {
+  	success = false;
+  }
+
+  *dispatch_mode = DispatchMode::kFComputeEx;
+
+  return success;
+}
+
+static bool CSRNeighborSampleShape(const nnvm::NodeAttrs& attrs,
+                                   std::vector<TShape> *in_attrs,
+                                   std::vector<TShape> *out_attrs) {
+  CHECK_EQ(in_attrs->size(), 3);
+  CHECK_EQ(out_attrs->size(), 1);
+
+  CHECK_EQ(in_attrs->at(0).ndim(), 2U);
+  CHECK_EQ(in_attrs->at(1).ndim(), 1U);
+  CHECK_EQ(in_attrs->at(2).ndim(), 1U);
+  // Check the graph shape
+  CHECK_EQ(in_attrs->at(0)[0], in_attrs->at(0)[1]);
+  // Probbality shape must be equal to the vertices length
+  CHECK_EQ(in_attrs->at(0)[0], in_attrs->at(1)[0]);
+
+  const NeighborSampleParam& params = 
+    nnvm::get<NeighborSampleParam>(attrs.parsed);
+
+  TShape out_shape(1);
+  out_shape[0] = params.max_num_vertices;
+  SHAPE_ASSIGN_CHECK(*out_attrs, 0, out_shape);
+
+  return out_attrs->at(0).ndim() != 0U &&
+         out_attrs->at(0).Size() != 0U;
+}
+
+static bool CSRNeighborSampleType(const nnvm::NodeAttrs& attrs,
+                                  std::vector<int> *in_attrs,
+                                  std::vector<int> *out_attrs) {
+  CHECK_EQ(in_attrs->size(), 3);
+  CHECK_EQ(out_attrs->size(), 1);
+  out_attrs->at(0) = in_attrs->at(0);
+
+  TYPE_ASSIGN_CHECK(*out_attrs, 0, in_attrs->at(2));
+  return out_attrs->at(0) != -1;
+}
+
+static void GetSample(std::vector<dgl_id_t>& ver_list,
+                      const float* probility,
+                      const size_t max_num_neighbor,
+                      std::vector<dgl_id_t>& out) {
+  // Copy ver_list to output
+  if (ver_list.size() <= max_num_neighbor) {
+    for (size_t i = 0; i < ver_list.size(); ++i) {
+      out.push_back(ver_list[i]);
+    }
+    return;
+  }
+  // Make sample
+  std::unordered_map<dgl_id_t, bool> mp;
+  while (out.size() < max_num_neighbor) {
+    random_shuffle(ver_list.begin(), ver_list.end());
+    for (size_t i = 0 ; i < ver_list.size(); ++i) {
+      int rand_num = (rand() % 100) + 1; 
+      float prob = probility[ver_list[i]];
+      if (rand_num <= static_cast<int>(prob*100) &&
+          mp[ver_list[i]] == false) {
+        out.push_back(ver_list[i]);
+        mp[ver_list[i]] = true;
+        if (out.size() >= max_num_neighbor) {
+          return;
```

Comment:
I don't think this sampling method actually samples vertices exactly following the given probability. The vertices at the beginning of ver_list have a higher chance of being selected. random_shuffle is also expensive.

An easier way is to compute the accumulative sum on the probability list and do a binary search on the accumulative probability list.

Reply:
binary search does require to remove the last sampled vertex and adjust the accumulative sum.
However, is this really mathematically equivalent? Could you show me how to calculate the probability a vertex is sampled? I'm not sure how to calculate when there is random_shuffle.